contracts/plugins/trading/DutchTrade.sol

```diff
@@ -183,6 +199,50 @@ contract DutchTrade is ITrade {
         bestPrice = _bestPrice; // gas-saver
     }
 
+    /// Bid with callback for the auction lot at the current price;
+    ///  Sold funds are sent back to the callee, callee.dutchTradeCallback(...) is invoked
+    ///  balance of buy token must increase by bidAmount(current block) after callback
+    ///  Trade is settled atomically via a callback
+    ///
+    /// @param data {bytes} The data to pass to the callback
+    /// @dev Caller must implement IDutchTradeCallee
+    /// @return amountIn {qBuyTok} The quantity of tokens the bidder paid
+    function bidWithCallback(bytes calldata data) external returns (uint256 amountIn) {
+        require(bidder == address(0), "bid already received");
+        // {buyTok/sellTok}
+        uint192 price = _price(block.number); // enforces auction ongoing
+
+        // {qBuyTok}
+        amountIn = _bidAmount(price);
+
+        // Mark bidder
+        bidder = msg.sender;
+        bidType = BidType.CALLBACK;
+
+        // status must begin OPEN
+        assert(status == TradeStatus.OPEN);
+
+        // reportViolation if auction cleared in geometric phase
+        if (price > bestPrice.mul(ONE_POINT_FIVE, CEIL)) {
+            broker.reportViolation();
+        }
+
+        // Transfer buy tokens from bidder
+        sell.safeTransfer(bidder, lot()); // {qSellTok}
+        uint256 balanceBefore = buy.balanceOf(address(this)); // {qBuyTok}
+        IDutchTradeCallee(bidder).dutchTradeCallback(address(buy), amountIn, data);
+        require(
+            amountIn <= buy.balanceOf(address(this)) - balanceBefore,
+            "insufficient buy tokens"
+        );
+
+        // settle() via callback
+        origin.settleTrade(sell);
+
+        // confirm callback succeeded
+        assert(status == TradeStatus.CLOSED);
+    }
+
     /// Bid for the auction lot at the current price; settling atomically via a callback
```

Comment:
This is now confusing, because the callback it is referring to is the `origin.settleTrade()` call. We should distinguish between the bidder callback and protocol callback in this contract.